solr/solrj/src/test/org/apache/solr/common/cloud/UrlSchemeTest.java

```diff
@@ -35,12 +35,12 @@ public void testApplyUrlScheme() {
     String liveNode1 = "192.168.1.1:8983_solr";
     String liveNode2 = "127.0.0.1:8983_solr";
     String liveNode3 = "127.0.0.1_";
-    String liveNode4 = "127.0.0.1:61631_l_%2Fig";
+    String liveNode4 = "127.0.0.1:61631_lx%2Fig";
```

Comment:
The changes to this test don't exercise the code change since they only have one underscore. It would be better to leave liveNode4 alone, and add, say, liveNode5 with a value like:
"127.0.0.1:61631_l_%2Fig_whatever"

Also see the comment in the JIRA.

Reply:
SOLR-15058 will break this UrlSchemeTest

Reply:
I think a good solution is to find the first underscore after the colon. If you look at `generateNodeName` in ZkController then you'll see the colon is always added to the generated `nodeName`. This may require some tests to update the test node name as there are some tests that use an improper node name that doesn't adhere to how `ZkController#generateNodeName` creates actual node names at runtime, this seems like a bad test impl. to me and should be corrected.

Reply:
fwiw ~ you can't look at the last underscore because nodeName encoding changes `/` to `_` in the context as you can't create znodes with `/` in them for obvious reasons ;-)

Reply:
Adding the following check to Utils:
```
diff --git a/solr/solrj/src/java/org/apache/solr/common/util/Utils.java b/solr/solrj/src/java/org/apache/solr/common/util/Utils.java
index 86bef9177c8..13e732d4ed7 100644
--- a/solr/solrj/src/java/org/apache/solr/common/util/Utils.java
+++ b/solr/solrj/src/java/org/apache/solr/common/util/Utils.java
@@ -759,7 +759,12 @@ public class Utils {
     return getBaseUrlForNodeName(nodeName, urlScheme, false);
   }
   public static String getBaseUrlForNodeName(final String nodeName, String urlScheme,  boolean isV2) {
-    final int _offset = nodeName.indexOf("_");
+    final int colonAt = nodeName.indexOf(':');
+    if (colonAt == -1) {
+      throw new IllegalArgumentException("nodeName does not contain expected ':' separator: " + nodeName);
+    }
+
+    final int _offset = nodeName.indexOf("_", colonAt);
     if (_offset < 0) {
       throw new IllegalArgumentException("nodeName does not contain expected '_' separator: " + nodeName);
     }
```

Results in the following test failures:
```
1669 tests completed, 3 failed, 20 skipped

4619 tests completed, 26 failed, 210 skipped

ERROR: The following test(s) have failed:
  - org.apache.solr.common.cloud.UrlSchemeTest.testApplyUrlScheme (:solr:solrj)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/solrj/build/test-results/test/outputs/OUTPUT-org.apache.solr.common.cloud.UrlSchemeTest.txt
    Reproduce with: gradlew :solr:solrj:test --tests "org.apache.solr.common.cloud.UrlSchemeTest.testApplyUrlScheme" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.client.solrj.routing.ShufflingReplicaListTransformerTest.testTransformReplicas (:solr:solrj)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/solrj/build/test-results/test/outputs/OUTPUT-org.apache.solr.client.solrj.routing.ShufflingReplicaListTransformerTest.txt
    Reproduce with: gradlew :solr:solrj:test --tests "org.apache.solr.client.solrj.routing.ShufflingReplicaListTransformerTest.testTransformReplicas" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.client.solrj.routing.ReplicaListTransformerTest.testTransform (:solr:solrj)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/solrj/build/test-results/test/outputs/OUTPUT-org.apache.solr.client.solrj.routing.ReplicaListTransformerTest.txt
    Reproduce with: gradlew :solr:solrj:test --tests "org.apache.solr.client.solrj.routing.ReplicaListTransformerTest.testTransform" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.handler.component.CloudReplicaSourceTest.testShardsParam_DownReplica (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.handler.component.CloudReplicaSourceTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.handler.component.CloudReplicaSourceTest.testShardsParam_DownReplica" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.handler.component.CloudReplicaSourceTest.testSimple_OnlyNrt (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.handler.component.CloudReplicaSourceTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.handler.component.CloudReplicaSourceTest.testSimple_OnlyNrt" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.handler.component.CloudReplicaSourceTest.testShardsParam_DeadNode (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.handler.component.CloudReplicaSourceTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.handler.component.CloudReplicaSourceTest.testShardsParam_DeadNode" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.handler.component.CloudReplicaSourceTest.testSimple_ShardsParam (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.handler.component.CloudReplicaSourceTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.handler.component.CloudReplicaSourceTest.testSimple_ShardsParam" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.handler.component.CloudReplicaSourceTest.testSimple_UsingClusterState (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.handler.component.CloudReplicaSourceTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.handler.component.CloudReplicaSourceTest.testSimple_UsingClusterState" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.handler.component.CloudReplicaSourceTest.testMultipleCollections (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.handler.component.CloudReplicaSourceTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.handler.component.CloudReplicaSourceTest.testMultipleCollections" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.handler.component.CloudReplicaSourceTest.testMultipleCollections_OnlyNrt (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.handler.component.CloudReplicaSourceTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.handler.component.CloudReplicaSourceTest.testMultipleCollections_OnlyNrt" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.handler.admin.HealthCheckHandlerTest.testFindUnhealthyCores (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.handler.admin.HealthCheckHandlerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.handler.admin.HealthCheckHandlerTest.testFindUnhealthyCores" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.core.CoreSorterTest.integrationTest (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.core.CoreSorterTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.core.CoreSorterTest.integrationTest" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cluster.events.ClusterEventProducerTest.testEvents (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cluster.events.ClusterEventProducerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cluster.events.ClusterEventProducerTest.testEvents" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.ClusterStateMockUtilTest.testBuildClusterState_Simple (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.ClusterStateMockUtilTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.ClusterStateMockUtilTest.testBuildClusterState_Simple" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.ClusterStateMockUtilTest.testBuildClusterState_ReplicaTypes (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.ClusterStateMockUtilTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.ClusterStateMockUtilTest.testBuildClusterState_ReplicaTypes" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.ClusterStateMockUtilTest.testBuildClusterState_ReplicaStateAndType (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.ClusterStateMockUtilTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.ClusterStateMockUtilTest.testBuildClusterState_ReplicaStateAndType" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverriddenZkACLAndCredentialsProvidersTest.testWrongCredentialsSolrZkClientFactoryUsingCompletelyNewProviders (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverriddenZkACLAndCredentialsProvidersTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverriddenZkACLAndCredentialsProvidersTest.testWrongCredentialsSolrZkClientFactoryUsingCompletelyNewProviders" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.NodeMutatorTest.downNodeReportsAllImpactedCollectionsAndNothingElse (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.NodeMutatorTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.NodeMutatorTest.downNodeReportsAllImpactedCollectionsAndNothingElse" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverseerTest.testRemovalOfLastReplica (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverseerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverseerTest.testRemovalOfLastReplica" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverseerTest.testExternalClusterStateChangeBehavior (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverseerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverseerTest.testExternalClusterStateChangeBehavior" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverseerTest.testOverseerFailure (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverseerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverseerTest.testOverseerFailure" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverseerTest.testStateChange (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverseerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverseerTest.testStateChange" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverseerTest.testShardLeaderChange (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverseerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverseerTest.testShardLeaderChange" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverseerTest.testShardAssignment (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverseerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverseerTest.testShardAssignment" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverseerTest.testDownNodeFailover (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverseerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverseerTest.testDownNodeFailover" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverseerTest.testReplay (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverseerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverseerTest.testReplay" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverseerTest.testOverseerStatsReset (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverseerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverseerTest.testOverseerStatsReset" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverseerTest.testBadQueueItem (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverseerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverseerTest.testBadQueueItem" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII

  - org.apache.solr.cloud.OverseerTest.testDoubleAssignment (:solr:core)
    Test output: /Users/tjp/dev/oss/lucene-solr/master/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.OverseerTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.OverseerTest.testDoubleAssignment" -Ptests.jvms=12 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=14FE50F5891679FC -Ptests.file.encoding=US-ASCII
```
This looks easy to fix up ... I'm out of the office (back Jan 5) so if you want to apply my update ^ and fix these tests (probably using improper nodeName values) @Sasasu ... otherwise, I'll fix and merge on Jan 5 ~ happy new year